### DISCUSSION TITLE
Sources transaction field: SUBMIT_POST

### DISCUSSION THREAD

**Discussion by HPaulson** [deso-protocol/dips#3] [Large CIPs]
---
CIP: To be assigned
Title: `Sources` field for `SUBMIT_POST` transactions
Author: Hunter Paulson <[@HPaulson](https://github.com/HPaulson)>
Discussions URL: https://github.com/bitclout/cips/discussions/3
Created: 2021-07-07
Note: This is a CIP duplication of [bitclout/core #32](https://github.com/bitclout/core/issues/32)

---

## One-Line Description

Add a new `Sources` field added to `SUBMIT_POST` transaction types to allow applications to specify where a post is meant to display.

## Explanation and Motivation

As the BitClout ecosystem expands, and many new applications are built on top of the protocol, having a standard for node-specific content will become vital. For example, if an engineer wanted to create an app built on top of the protocol which only displayed posts meant for that app (For example, if you wanted to make a Reddit-like app, it wouldn't make sense to display all posts from all subreddits on one twitter-like feed), the app could formulate a field like such in the request: `Sources: ["myApp.com"]`. BitClout.com, and any other nodes looking to display all content, could use the `"Default"` source to signify it displays regular content not specific to an application type. Those nodes could filter by this source. This could allow engineers to display posts only allowed for their source, without doing anything too hacky.

I think for the long-term success of BitClout we need many types of applications built on top of the protocol, and additions like this would be very very helpful in accomplishing such.

### Backend
The new field, `Sources: []string`, should be added to typings for  both [`PostEntryResponse`](https://github.com/bitclout/backend/blob/a20d4e311b2913f7d0b9e46ed72c82830dc83889/routes/post.go#L56) and [`SubmitPostRequest`](https://github.com/bitclout/backend/blob/d4b3ccd11d6ce27ef82d93c3a20facd8efd44c0f/routes/transaction.go#L1069). Additionally, the [`SubmitPost`](https://github.com/bitclout/backend/blob/d4b3ccd11d6ce27ef82d93c3a20facd8efd44c0f/routes/transaction.go#L1108) function should be updated to include the data from `Sources` in its call to [`CreateSubmitPostTxn`](https://github.com/bitclout/backend/blob/d4b3ccd11d6ce27ef82d93c3a20facd8efd44c0f/routes/transaction.go#L1255).

In addition, the backend should add a conditional to check if a post is meant for the node it's running on -- though it would be at the discretion of the node operator to set such using a comma-separated list of the accepted sources in a backend flag. For example, `--sources "prism.red, love4src.com"` would active such a conditional to only return posts which include `prism.red` OR `love4src.com` in its `Sources` array, and no flag would include all posts regardless of its `Sources`. BitClout.com's node should be updated accordingly, displaying posts that include the `Default` source.

### Core
The [`CreateSubmitPostTxn`](https://github.com/bitclout/core/blob/4421c7e080c102e0dd56b54ec80feb618f14775d/lib/blockchain.go#L2593) function should be updated to accept the `Sources` data as a parameter. The `Source` field should also be added to the [`SubmitPostMetadata`](https://github.com/bitclout/core/blob/4421c7e080c102e0dd56b54ec80feb618f14775d/lib/network.go#L3190) struct with its appropriate `[]byte` type. Finally, the usages of `SubmitPostMetadata` should be updated: [Main](https://github.com/bitclout/core/blob/4421c7e080c102e0dd56b54ec80feb618f14775d/lib/blockchain.go#L2628), [Test](https://github.com/bitclout/core/blob/135c03a958039423ac2c775cb83eb2a41d903511/lib/network_test.go#L778)

### Frontend
The [`SubmitPost`](https://github.com/bitclout/frontend/blob/7cfb241668518beb8f22a4af6e520aa5926129d9/src/app/backend-api.service.ts#L659) method should be updated to include `Sources: Array<string>` as an parameter, and should include such as an argument when calling the [`Api post function`](https://github.com/bitclout/frontend/blob/7cfb241668518beb8f22a4af6e520aa5926129d9/src/app/backend-api.service.ts#L672). Additionally, the [`submitPost`](https://github.com/bitclout/frontend/blob/7403ded2b65000833bc7be45781f3eea3bfb388f/src/app/feed/feed-create-post/feed-create-post.component.ts#L130) method should be updated accordingly, including `Sources: ["Default"]` in it's [call](https://github.com/bitclout/frontend/blob/7403ded2b65000833bc7be45781f3eea3bfb388f/src/app/feed/feed-create-post/feed-create-post.component.ts#L160) to `SubmitPost()`.

## Other Solutions
While methods for accomplishing such a task already exist, for example hiding a post by default and making use of `PostExtraData` to identify the source application & if the post is *actually* hidden or not, however, I think having a standard method of doing such would be a much better, and long-lasting, solution.

## Backwards Compatibility
There are no known issues with backward compatibility.

## Test Cases
The only required test should be an update to [`TestSerializeSubmitPost`](https://github.com/bitclout/core/blob/135c03a958039423ac2c775cb83eb2a41d903511/lib/network_test.go#L772), as described above.

## Security Considerations
No known concerns.

**Comment by connerdouglass** [comment]
Hey @HPaulson, congrats for starting the first community proposal discussion in this repo! Overall this is an interesting idea. I have a few thoughts that I think are worth exploring:

- Why call it `Sources`? Because it's just a slice of strings added to the transaction meta, I think they're more like `Tags`.
- If you want it to be cryptographically enforceable, nodes need to have a public key associated with them, and they need to use the associated private key to affix a signature to every post originating from their node. I don't support this, just wanted to mention it.

My thinking currently is that adding a slice of strings to the transaction wouldn't achieve the desired filtering goals, but I would really like to hear more about this to clear up my understanding!

**Reply by HPaulson** [reply]
1. Originally my idea was to have it be one string, the "Source" node where the post came from, but realized having multiple makes much much more sense -- this was my rational, but quite honestly, the naming really doesn't matter.

2. I disagree entirely on this point. The purpose of the sources is to designate that a post is "supposed to" display on a particular node -- the goal is not to make it so only that node can add its designator to a post, or to make it enforceable in any way. For my specific use case, as an example: Prism is an application which is Instagram-like in nature. Considering this, having it share all posts with an application which is Twitter-like wouldn't make much sense, as often images on Instagram are much more high quality that those you may post on Twitter. For this purpose, posts on Prism won't (by default) display on other nodes, and posts from other nodes won't display on Prism, as having such wouldn't make sense. So, having a specified tag or source which confirms the post is meant for Prism makes a ton of sense -- and is more viable than using the hacky method we rely on now (see "Other Solutions"). It also allows for a more unified spec on the topic, as I'm aware of other applications looking to do the same. Anyone could add their post to Prism even while not using the app itself, by adding such a designator, so no need to make adding the source exclusive. The goal isn't to restrict who can add a post to what app, but to clarify that a post SHOULD display on a particular app. (Unsure if I explained this great, it's super late. Let me know if you need clarification!)

**Reply by bloated-devish** [reply]
Few questions:

1) I'd strongly suggest that you mention the size of this ```Sources``` backend field. What is the size of each string? How many different strings can be added? 

2) I personally would prefer an integer enum so it uses less space (and possible to retain with a bitset array for multiple sources). This is especially relevant, since a not-so forward looking application can start making posts costlier by simply having large tags. 

3) Do you care about adding a core / backend lookup of posts meeting specific source criteria? This might be useful for frontend nodes.

Also food for thought: Are you expecting this to grow into something more inter operable? Like tomorrow would this be something that a class of applications could interoperate? Do you want all prism type apps to consume this post. In that case, we could have the enum be grouped - assign some range for apps of various types and create a util func for group identification.

Thanks -Bloated

**Reply by HPaulson** [reply]
Sorry for the late reply here. 

1. I didn't consider any specific size. Here's why: Since the body of a post (Both the content, and imageURLs array) is already "Unlimited" in size (Only restricted by transaction size to my knowledge), I assumed this would be no different 
2. I'm not exactly sure what specifically you're suggesting. Could you please clarify?
3. I mention this is the "Backend" section of the specification

As far as food thought -- 100%. The intention is to allow groups of apps to work together and allow for cross-support between them. For example, if Application A wanted to display posts from Application B and C, it could make the appropriate implementations according to those app's spec (Likely has special data in PostExtraData), and then include such applications in its Source flag backend. This will allow for collaboration, but make it so apps aren't grouped -- as with grouping, important post data in PostExtraData may be ignored. This ensures the full value of posts is included while collaborating among projects, rather than being missed by accepting posts from a large group of apps. The only "Group" source I'd recommend is the Default group, mentioned above.

**Comment by dgsus** [comment]
i believe its important to be able to pinpoint where a transaction came from, apart from the use case Hunter mentioned, i see a at least one other good case -> forensic investigations of bad actors

not sure if the best thing would be a public key associated with them, or some sort of unique hash based on few properties of the node (physical/virtual , network, etc)

**Reply by tijno** [reply]
These are probably separate topics.

The "sources" or "tags" for a post are helpful for discovery and nodes wanting to display specific content.

But a "which node submitted this transaction" feature i think is very different, has all sorts of privacy and decentralisation implications, and may already have been addressed in the code (although i could not see it).

**Reply by dgsus** [reply]
indeed, it seems my mind just ignored the part that this was just for submit_post , thx for pointing it out.

let me just wrap this up just by saying:
maybe i wasnt very clear but i meant in a way we could cluster together transactions from a node by - lets say- node-id. this would be a pseudonymous id, we should not be able to tell where is the node directly from this id (no ip or such info) 

maybe in the code already as you mentioned...

**Reply by tijno** [reply]
I think it would be very useful as a separate discussion  - not necessarily for implementation right now - but its a complex topic that needs some good quality thought & time to mature. So lets get that discussion started?

**Reply by HPaulson** [reply]
@Dgsus Yeah I think this is an amazing idea if done correctly (as Tijn mentioned, lot of privacy concerns), but is a separate topic. I'd love to see a separate Discussion created on this!

**Reply by dgsus** [reply]
Yes will submit something later tonight to start this discussion

**Reply by bloated-devish** [reply]
I have another concern. Could someone actually create tons of PRISM tagged posts, but just useless, just to make the PRISM app slow? In such a situation, would it be relevant to actually PRISM identify itself - that this is from the PRISM app. The solution there could be to encrypt the transaction/content or even some specific field with the node's private key and let the trust block provider decrypt using the node's public key to confirm and tag it as prism. This is a little complicated idea (while not sure if there is an underlying problem to solve), with some security implications - the node's key is in the config, but should also not get leaked (config file github submits are not unheard of), but can solve the non repudiation of node problem. Privacy implications not any different from current proposal.

**Reply by HPaulson** [reply]
@bloated-devish This isn't really relevant as it's already possible today. Someone could easily do the exact same thing without source tags, no change. Again, the idea here is not to have a bunch of proprietary posts -- that'd defeat the purpose of bitclout. Rather it's to allow engineers to communicate easily with nodes that a post authed by the user is meant for that node. This is it. Anything beyond this is entering either dangerous regulatory/decentralization territory, or is overly complex and solves a non-existent problem. e

**Comment by HPaulson** [comment]
@lazynina What exactly does "the community [...approving] your proposal" consist of? Is the above discussion adequate to request core team approval? If not, what further discussion shall occur? Is there an ideal time period a discussion should be open for before becoming a PR? (Asking more so in general, but applies to this discussion)

**Reply by lazynina** [reply]
@HPaulson I don't have an exact definition of "the community [...approving] your proposal" at this time. My hope is that it becomes more defined after CIPs gets rolling for a bit, but for now we don't want process to hold up progress.

Generally, I think a discussion should be open for at least a few days (unless it is a trivial CIP) before drafting a PR.  This allows both core devs and community members ample time to provide their thoughts on the proposal. It also helps save the CIP drafter time by reducing revisions of the CIP PR. 

In this instance, I think there is more discussion to be had - I've got a few questions/concerns that I need to type up :sweat_smile:

**Reply by tijno** [reply]
> for now we don't want process to hold up progress

nina not lazy - keep pushing.

(i owe you about 20 PR reviews)

**Reply by HPaulson** [reply]
@lazynina Sounds great! Thanks for clarifying :)

**Comment by lazynina** [comment]
Thanks for the great discussion. Bit of a brain dump here -- I think there's a way to make this work, but below are my concerns/recommendations/questions.

1. Adding this to the SubmitPostMetadata would create a fork as it would change how SubmitPostMetadata is serialized.  The ExtraData field of transactions is meant to support these non-core attributes and is where the Source key should live in the transactions instead of the SubmitPostMetadata.  ExtraData is effectively an arbitrary map to store data about a transaction.
2. The sources attribute could become really bloated in the future when there are 1000s of nodes. If I want my post to show up on all these nodes, I'm going to need to include the sources filtered on for each of those nodes. This would lead to higher fees and generally inefficient storage of posts.
3. Existing posts and new posts without this source attribute would not pass the filters of these other nodes and in a way leads to unintentional censoring of a post.  This becomes confusing as a user -- I posted on bitclout.com, but my post will not appear on prism or another node and it's not clear to the lay user why this happens.
4. Any post transaction could set whatever it wants as the source (as mentioned above) so I could have a post with no image set to display on prism. 
5. How does source work in comments? Do I need to have the same source as the parent in order to show up on the prism node? Or does the filtering only apply to the root level post? Just curious what the thoughts are on this.

Perhaps instead of source, we can have community agreed upon tags, which are a bit more general - e.g. "high-quality image", "nsfw", etc. However, the categorization of posts may not be efficient to maintain on-chain in the long run.

*brain dump over - rebooting nina*

**Reply by HPaulson** [reply]
These are some great points, but I think most have a fairly straightforward answer:

1. This is totally fine, and likely makes implementation easier anyway. In this case, this CIP becomes more a spec definition than an actual transaction change -- which lgtm.
2. This is why the "Default" source mentioned above would exist. In reality, thousands of nodes wouldn't need much outside of the default scope. Say, for example, someone wanted to make use of Prism's structure through a CLI -- all they'd have to do is use the Prism flag, and have no worries about bloating it otherwise. I can't think of a situation at scale where there are many nodes all using each other as a source, outside of the Default source group (As defined in the CIP).
3. This is an amazing point -- and one I didn't consider. I still think, however, the answer is fairly simple: Since all posts on the chain now have the properties of the Default source group (This is by nature, as since no other sources exist, and the Default group is a replication of what we have currently, all current posts would fit into this category). I think as such, any post without a source could likely be defined (by default) as a Default post. So, the "Default" group would be both "Default" in the Sources array, OR no sources array at all (Or an empty one). In the very rare scenario which this isn't the case (For example, the post we made with Prism using our current workaround) could just be updated very easily, as I can't think of really any scenarios outside of the previous (And other apps with similar -- which tmk don't exist outside of testing at the moment) where this exists. If this assumption is incorrect, I'd love to hear from someone who implemented such a post where this assumption does not apply -- I think this is very very rare. 
4. Of course! This is a feature -- not a bug. By enforcing this in some way, there's a whole list of issues that arise -- and overall isn't a good idea. The purpose of such is to simply have a flag that designates that X node should or should not display -- but it's still the responsibility of the node to perforce adequate checks. The source should be used as a tool, not as something to rely on. In this case, provided, We'd check for posts that are meant for prism, and also have an image (Also checks if said image is valid in our network, if not it will display "Image unavailable" but still show the post).
5. Another amazing point which I hadn't considered. If sources are implemented, I think it'd be the case that only the root matters as comments would be handled by node accordingly (For example in our use case, we'd display all comments but omit images in them). I do think, however, comments could easily also utilize source flags, as the node could just filter all posts. Quite honestly I don't think it makes a major difference.

I don't think categorizing in this way makes much sense, to be honest. Curious about more perspectives here -- for me personally, I think this could easily break use cases due to subjectivity. I think the Default group covers about 90% of use cases, where the extra 10% can rely on sources.

Thanks a ton for this -- I didn't consider these last two points, and they're very valid concerns. Great for thought food!

**Reply by bloated-devish** [reply]
Based on @lazynina 's comment, I am convinced that these sources will explode - and its entirely possible that a post can have multiple 'tags' (nsfw and high quality image are not mutually exclusive). Even in ```ExtraData```, should we consider these as enums? However, as enums, we need to make sure the process to add new enum value less painful for everyone.

-Bloated

**Reply by connerdouglass** [reply]
@bloated-devish I agree with that, and I think an elegant solution would be to encode these tags into txns as a single uint32 bitmask, where each bit position has some accepted meaning.

Plus since it would be a single uint32, it's extensible as the community desires more tag options (up to 32 of them).

Thoughts on this @lazynina?

**Reply by HPaulson** [reply]
@lazynina Also one mention here -- Since the PostExtraData field only accepts string types, I'm unsure how you expect this to work (Which is why I initially suggested it be in metadata). Since this would need to be an array of data, and all PostExtraData values must be strings, I'm unsure of your expectation here.

**Reply by connerdouglass** [reply]
@HPaulson if it helps, I'm pretty sure PostExtraData is a `map[string][]byte` type so it can contain any data, as long as the data is first encoded to a slice of bytes. An array of strings can be encoded as a byte slice pretty easily, so I think your design is fine on those grounds!

**Reply by HPaulson** [reply]
Ah okay good shot

**Reply by HPaulson** [reply]
@lazynina Any response to my comments above? How should we proceed further?

**Reply by lazynina** [reply]
> @lazynina Also one mention here -- Since the PostExtraData field only accepts string types, I'm unsure how you expect this to work (Which is why I initially suggested it be in metadata). Since this would need to be an array of data, and all PostExtraData values must be strings, I'm unsure of your expectation here.

I may not have been clear about the distinction between the PostExtraData attribute of the SubmitPostRequest and the ExtraData attribute of a transaction. Adding a new attribute to the SubmitPostRequest struct to support “sources” or “tags” would be a smaller change, but adding it to the SubmitPostMetadata instead of the ExtraData attribute of the transaction would cause an avoidable fork as the serialization of the SubmitPostMetadata would change.

The backend would merge the source into the extra data when creating the transaction. The PostExtraData attribute of the SubmitPostMetadata is just a convenience for future addition of simple string-to-string key-value mappings that can live in the transaction’s ExtraData.  You can see how we can handle merging other attributes into the extra data map here as we do for reclout keys: https://github.com/bitclout/core/blob/main/lib/blockchain.go#L2616

**Reply by lazynina** [reply]
@maebeam Any thoughts on this CIP and what the best way to allow nodes to filter without blowing up the transaction size for posts in the long term?

**Reply by HPaulson** [reply]
@maebeam Unsure if ya saw this mention or not

**Reply by maebeam** [reply]
Apologies for the delay.

I think a source field for the purposes of displaying which node a post came from and using it to direct traffic to other nodes could be cool. My biggest concern is storage efficiency and I am actually pro adding a field to SubmitPostMetadata that is an integer enum of sources. (Side note: I want to have a longer discussion on a different thread about extra data. I'm worried about it blowing up the chain size.) This approach uses storage efficiently, encourages node operators to contribute to the codebase, and we could require at least some community + core dev threshold for adding nodes to the list of sources. I imagine this being defined in code as a map from integer to Node Name, Node URL. Hunter wants to use this field to filter posts which I am fine with. 

Regarding tagging of posts, I don't think tags belong on chain. A tag would be permanent or only changeable by the post author. Realistically I see someone running a tags service that employs machine learning to provide tag metadata for posts on the chain. This solves the backfill problem and acknowledges that tags change overtime as models evolve and new tags are created.

**Reply by HPaulson** [reply]
Yeah I definitely support this model. Initially I thought the int enum was only a suggestion for a tag system, which I don’t think is a great idea (as you to touched on). Using it to define a node makes sense to me though, my only concern is a node getting issued an int and then dying/closing. How could this be addressed? If it’s not an issue. What do our next steps look like? @maebeam

**Reply by tijno** [reply]
This is a good solution.

Re nodes dying - the entry stays as the content remains on chain but maybe we add something to the name and change the url for the node to point to post confirming node shutdown ? 

Eg:

15
nodename [dead]
https://docs.bitclout.com/nodes/graveyard

**Reply by ryanjohn1** [reply]
Thanks @HPaulson for raising this, the question of how apps/nodes/services build on top of bitclout and add their own data and deviations from what has been done so far is going to be really important to get right. It's something we've begun considering in earnest at Flick as we're getting started on our first major features that are off in our own direction rather than just creating a mobile version of Bitclout.com. That's going to be happening more and more now as different projects reach that stage and forge their own path, that diversifying is a really important stage for the platform but has lot's of challenges too.
As @maebeam says, I definitely think we should have a wide ranging discussion on post extra data, best practices and any improvements we might need to avoid inflating the chain size. Is there a discussion on the go for this yet?

**Reply by HPaulson** [reply]
I agree, we definitely need some spec for extra data. I'll let someone else handle that thread elsewhere, though.

For now, I think @tijno brings up a good point, and I see a clear path forward -- though I'm hoping this updating of node declaration isn't too much maintenance

**Reply by bloated-devish** [reply]
HI @HPaulson @ryanjohn1 @tijno 

Can you have a look at this proposal I have: https://github.com/bitclout/cips/discussions/49#discussioncomment-1020381 in the context of ```ExtraData```. Let me know what you think there. 


Thanks\
Bloated

**Reply by HPaulson** [reply]
@lazynina @maebeam How can we proceed further?

**Reply by HPaulson** [reply]
Still waiting on this, has been well over a month now. @maebeam @lazynina

**Reply by maebeam** [reply]
This was implemented recently! Marking DIP as approved.

**Reply by HPaulson** [reply]
@maebeam The other section of this, which is equally important IMO, is being able to serve posts only from X node. Any plans on implementing this? @maebeam

**Reply by maebeam** [reply]
While I'd love to implement that feature I don't have any immediate plans to. Feel free to implement it though, I would happily review and merge a PR.
